src/rShare.ts

```diff
@@ -0,0 +1,322 @@
+import * as vscode from 'vscode';
+import * as vsls from 'vsls';
+import * as fs from 'fs-extra';
+
+import { enableSessionWatcher, extensionContext } from './extension';
+import { attachActiveGuest, browserDisposables, initGuest, removeGuestFiles } from './rShareSession';
+import { initTreeView, rLiveShareProvider, ToggleNode } from './rShareTree';
+import { Commands, Callback, onRequest, request } from './rShareCommands';
+import { HelpFile } from './rHelp';
+
+/// LiveShare
+export let rHostService: HostService = undefined;
+export let rGuestService: GuestService = undefined;
+export let liveSession: vsls.LiveShare;
+export let isGuestSession: boolean;
+export let _sessionStatusBarItem: vscode.StatusBarItem;
+const disposables: vscode.Disposable[] = [];
+
+// service vars
+export const ShareProviderName = 'vscode-r';
+export let service: vsls.SharedServiceProxy | vsls.SharedService | null = undefined;
+
+// random number to fake a UUID for differentiating between
+// host calls and guest calls (specifically for the workspace
+// viewer 'View' function)
+export const UUID = Math.floor(Math.random() * Date.now());
+
+/// state-tracking bools
+// Bool to check if live share is loaded and active
+export async function isLiveShare(): Promise<boolean> {
+    const shareStarted = (await vsls.getApi())?.session.id;
```

Comment:
When I test on macOS, this line will take forever and the extension will not be fully activated. Not sure what caused this while it does not seem to happen on Ubuntu.

Reply:
Very weird, I don't get this on WSL either... I'll try to get my hands on a mac to see if I can reproduce

Reply:
Is there anything we could improve by learning how vscode-jupyter does this?

https://github.com/microsoft/vscode-jupyter/blob/b1f70d0fc436c9bade9941406b2590b4034cf2bf/src/client/datascience/liveshare/liveshare.ts#L40-L62

Reply:
If I understood the code correctly, this command tries to activate the live share extension/api when it's called, but the description (and function name) suggest that it only checks if the api is active. Would it make sense to split this into two separate functions (e.g. `isLiveShare` and `activateLiveShare`)? Because if the extension isn't active, there can't be an active session anyways, and then we'd cause unnecessary overhead by activating the live share extension.

Reply:
Thanks for checking out the PR @ManuelHentschel! The code is *intended* to check if there is an active live share session.

`vsls.getApi()).session.id` doesn't (or shouldn't) start a live share session on its own, it should be just returning the api for the live share SDK, which contains information on the current session. If liveshare is installed, `vsls.getApi()` will always return TRUE, which isn't especially useful - hence checking if there is a session id.

I've tried to pinpoint how the liveshare extension activates (though this is a little difficult given it being closed-source). From what I can tell, the extension is always active when installed (and enabled), so the overhead presumably wouldn't be any different to the user having the extension enabled. Admittedly, I'm not sure how to track extension performance...

Having said that, it isn't especially useful to grab the api twice, so I've changed it to rely on the liveSession variable, which is set during `initLiveShare`.

Thank you again!

---

@renkun-ken Hi Ken, thanks again for having a look at the PR. I had a look at the jupyter code, but I was a little confused as to why they re-implement the vsls API. It seems like they've done so to implement some custom commands, but otherwise the functionality seems to be identical to the vsls API. Am I wrong in this understanding?

I've based the PR on a few different liveshare repos:

- https://github.com/vsls-contrib/togezr/blob/7d814f1182efea6dd3d44945423a701e9e34a802/vscode-extension/src/branchBroadcast/liveshare.ts#L33
- https://github.com/vsls-contrib/pairit/blob/master/src/extension.ts#L41-L63
- https://github.com/lostintangent/whiteboard/blob/master/src/service/guestService.ts

Reply:
> I've tried to pinpoint how the liveshare extension activates


On the extension page you can check Feature Contributions > Activation Events to see when an extension activates. In this case, the list includes `*`, which means that the extension will always be activated on startup, so afaik you don't need to worry about accidentally activating the extension yourself.

Reply:
Let me try again and see if it works.